FlaSpaceInst/EZ-RASSOR
# Add visual odometry and integrate with Kalman filter

# Description
This PR includes the following changes:

- Add the `rgbd_odometry` node from `rtabmap_ros` for visual odometry.
- Add the output from the `rgbd_odometry` node to the Kalman filter.
- Remove the `transform_broadcaster` node. This was used in the first iteration of EZ-RASSOR to publish the transform for the left lens of the stereo camera for visual odometry, so it is no longer necessary. For visual odometry using a depth camera, a link `depth_camera_optical_frame` has been added.

# Testing
- Run `sh develop.sh resolve` to install the added dependencies (`rtabmap_ros`).
- Run the following to spawn the robot in the moon world (the default world does not have visual features for visual odometry to use) in autonomous mode with odometry enabled.
```
roslaunch ezrassor_launcher configurable_simulation.launch control_methods:="autonomy" enable_real_odometry:="true" world:=moon digsite_x_coords:=-10 digsite_y_coords:=-10 spawn_x_coords:=2 spawn_y_coords:=-2
```
- In a separate terminal, echo the visual odometry topic to monitor what it's publishing.
```
rostopic echo /ezrassor1/visual_odom/odom
```
- In another separate terminal, enable auto-drive.
```
rostopic pub /ezrassor1/autonomous_toggles std_msgs/Int8 1
```
- The robot should now go towards (-8, -12) since the digsite coordinates are relative to the robot's starting position. Once it has reached what it thinks is its destination, it will stop and log "Destination reached!" On the left side of the Gazebo screen, click Models -> ezrassor1 -> pose to see how far off the robot was from actually reaching (-8, -12). The x and y values here should be pretty close to -8 and -12, respectively.
